### PR TITLE
Shows upcoming events on public events page

### LINKED ISSUE #926: Only Show Future Events on Event Page
When users come to Core-v4, they should see a list of upcoming events on the events page. The page calls the `/getEvents` API which is defined in [`api/main_endpoints/routes/Event.js`](https://github.com/SCE-Development/Core-v4/blob/abf521d58e1ee6becb93ef165e334614638bff55/api/main_endpoints/routes/Event.js#L18). We need to update the query to only retrieve the events that occur after todays date. Furthermore, on the Event admin page, all events should be returned regardless of date.

Small mongoDB change.

### PR DESCRIPTION
**Pull request by neelp03**
### Resolves #926 
- created API that retrieves all upcoming events (including today's)
- created API function for `events/getUpcomingEvents`
- updated public event page to populate the page with all upcoming events by calling `getUpcomingEvents()` api function

> Note: Public view shows only upcoming events while admin view (event manager) shows all events as seen below (look at the dates).

### CODE REVIEW

File: api/main_endpoints/routes/Event.js [hunk 1]
```diff
@@ -3,27 +3,43 @@ const router = express.Router();
...
-const {
-  OK,
-  BAD_REQUEST,
-  UNAUTHORIZED,
-  FORBIDDEN,
-  NOT_FOUND
-} = require('../../util/constants').STATUS_CODES;
-const addErrorLog = require ('../util/logging-helpers');
+const { OK, BAD_REQUEST, UNAUTHORIZED, FORBIDDEN, NOT_FOUND } =
+  require('../../util/constants').STATUS_CODES;
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:9]:
```suggestion
const {
  OK,
  BAD_REQUEST,
  UNAUTHORIZED,
  FORBIDDEN,
  NOT_FOUND
} = require('../../util/constants').STATUS_CODES;
```

File: api/main_endpoints/routes/Event.js [hunk 2]
```diff
@@ -3,27 +3,43 @@ const router = express.Router();
...
+    },
+  })
+    .sort({ eventDate: -1, startTime: -1 }) // Sort By date in descending order
+    .then((items) => res.status(OK).send(items))
+    .catch((error) => {
       const info = {
         errorTime: new Date(),
         apiEndpoint: 'Event/getEvents',
-        errorDescription: error
+        errorDescription: error,
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:42]:
```suggestion
        errorDescription: error
```

File: api/main_endpoints/routes/Event.js [hunk 3]
```diff
@@ -44,7 +60,7 @@ router.post('/createEvent', (req, res) => {
     startTime: req.body.startTime,
     endTime: req.body.endTime,
     eventCategory: req.body.eventCategory,
-    imageURL: req.body.imageURL
+    imageURL: req.body.imageURL,
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:63]:
```suggestion
    imageURL: req.body.imageURL
```

File: api/main_endpoints/routes/Event.js [hunk 4]
```diff
@@ -69,10 +85,10 @@ router.post('/editEvent', (req, res) => {
     startTime,
     endTime,
     eventCategory,
-    imageURL
+    imageURL,
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:88]:
```suggestion
    imageURL
```

File: api/main_endpoints/routes/Event.js [hunk 5]
```diff
@@ -69,10 +85,10 @@ router.post('/editEvent', (req, res) => {
...
   } = req.body;
   Event.findOne({ _id: req.body.id })
-    .then(event => {
+    .then((event) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:91]:
```suggestion
    .then(event => {
```

File: api/main_endpoints/routes/Event.js [hunk 6]
```diff
@@ -83,17 +99,17 @@ router.post('/editEvent', (req, res) => {
       event.imageURL = imageURL || event.imageURL;
       event
         .save()
-        .then(ret => {
+        .then((ret) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:102]:
```suggestion
        .then(ret => {
```

File: api/main_endpoints/routes/Event.js [hunk 7]
```diff
@@ -83,17 +99,17 @@ router.post('/editEvent', (req, res) => {
...
           res.status(OK).json({ ret, event: 'event updated successfully' });
         })
-        .catch(error => {
+        .catch((error) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:105]:
```suggestion
        .catch(error => {
```

File: api/main_endpoints/routes/Event.js [hunk 8]
```diff
@@ -83,17 +99,17 @@ router.post('/editEvent', (req, res) => {
...
           res.status(BAD_REQUEST).send({
             error,
-            message: 'event was not updated'
+            message: 'event was not updated',
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:108]:
```suggestion
            message: 'event was not updated'
```

File: api/main_endpoints/routes/Event.js [hunk 9]
```diff
@@ -83,17 +99,17 @@ router.post('/editEvent', (req, res) => {
...
           });
         });
     })
-    .catch(error => {
+    .catch((error) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:112]:
```suggestion
    .catch(error => {
```

File: api/main_endpoints/routes/Event.js [hunk 10]
```diff
@@ -105,10 +121,10 @@ router.post('/deleteEvent', (req, res) => {
     return res.sendStatus(UNAUTHORIZED);
   }
   Event.deleteOne({ _id: req.body.id })
-    .then(event => {
+    .then((event) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:124]:
```suggestion
    .then(event => {
```

File: api/main_endpoints/routes/Event.js [hunk 11]
```diff
@@ -105,10 +121,10 @@ router.post('/deleteEvent', (req, res) => {
...
       res.status(OK).json({ event: 'event successfully deleted' });
     })
-    .catch(error => {
+    .catch((error) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:127]:
```suggestion
    .catch(error => {
```

File: test/Event.js [hunk 12]
```diff
@@ -206,4 +228,35 @@ describe('Event', () => {
       expect(getEventsResponse).to.have.length(0);
     });
   });
+  describe('/GET getUpcomingEvents', () => {
+    it('Should return 404 if no upcoming events are available', async () => {
```
Comment (evanugarte) [inline test/Event.js:232]:
```suggestion
    it('Should return an empty array if no upcoming events are available', async () => {
```

File: test/Event.js [hunk 13]
```diff
@@ -206,4 +228,35 @@ describe('Event', () => {
...
+      setTokenStatus(true);
+      const result = await test.sendGetRequest(
+        '/api/event/getUpcomingEvents');
+      expect(result).to.have.status(OK);
+      const getEventsResponse = result.body;
+      getEventsResponse.should.be.a('array');
+      expect(getEventsResponse).to.have.length(0);
+    });
+    it('Should return 200 if there are upcoming events', async () => {
```
Comment (evanugarte) [inline test/Event.js:241]:
```suggestion
    it('Should return all upcoming events', async () => {
```

File: api/main_endpoints/routes/Event.js [hunk 14]
```diff
@@ -12,14 +12,35 @@ const {
   FORBIDDEN,
   NOT_FOUND
 } = require('../../util/constants').STATUS_CODES;
-const addErrorLog = require ('../util/logging-helpers');
+const addErrorLog = require('../util/logging-helpers');
 const membershipState = require('../../util/constants').MEMBERSHIP_STATE;
 
 router.get('/getEvents', (req, res) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:18]: the api handlers `/getEvents` and `/getUpcomingEvents` effectively do the same thing with a minor difference.

to save code we can abstract it to a function

something like:
```js
function eventQueryHandler(res, date=false) {
    let maybeQuery = {}
  if (date) {
    // insert logic here for calculating today
    maybeQuery = {
      eventDate: {
        $gte: today.toISOString(), // gets all upcoming events including today's
      },
    };
  }
  Event.find(maybeQuery)
      // paste the rest of the query here, there the same in both handler

}
```

`date` specifies if we want events today and beyond, res is the response to the user. the two functions become:

```js
router.get('/getEvents', (req, res) => {
  getAllEvents(res);
});

router.get('/getUpcomingEvents', (req, res) => {
  getAllEvents(res, true);
});
```

**Review by evanugarte — APPROVED** [review]
Check the one suggestion, amazing work!

File: api/main_endpoints/routes/Event.js (hunk 14 above)
Comment (evanugarte) [reply api/main_endpoints/routes/Event.js:18]: with this change the behavior should stay exactly the same and we cut down on code

Comment (neelp03) [reply api/main_endpoints/routes/Event.js]: I simplified the two get event requests like this: https://github.com/SCE-Development/Core-v4/pull/933/commits/b841b5a537ca1baae2402ffee63ba366e9ed9279
> fixed syntax issues in the next 2 commits

```js
function getEvents(req, res, upcomingEvents) {
  let query = {};
  if(upcomingEvents) {
    const today = new Date(Date.now());
    today.setHours(0, 0, 0, 0);
    query = {
      eventDate: { $gte: today.toISOString() }
    };
  }
  Event.find(query)
    .sort({ eventDate: -1, startTime: -1 })
    .then((items) => res.status(OK).send(items))
    .catch((error) => {
      const info = {
        errorTime: new Date(),
        apiEndpoint: 'Event/getEvents',
        errorDescription: error,
      };
      addErrorLog(info);
      res.status(BAD_REQUEST).send({ error, message: 'Getting event failed' });
    });
}

router.get('/getEvents', (req, res) => {
  getEvents(req, res, false);
});

router.get('/getUpcomingEvents', (req, res) => {
  getEvents(req, res, true);
});
```

File: api/main_endpoints/routes/Event.js [hunk 15]
```diff
@@ -12,22 +12,39 @@ const {
   FORBIDDEN,
   NOT_FOUND
 } = require('../../util/constants').STATUS_CODES;
-const addErrorLog = require ('../util/logging-helpers');
+const addErrorLog = require('../util/logging-helpers');
 const membershipState = require('../../util/constants').MEMBERSHIP_STATE;
 
-router.get('/getEvents', (req, res) => {
-  Event.find()
+function getEvents(req, res, upcomingEvents) {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Event.js:18]: i don't believe req is used in this function. pls remove

Comment (evanugarte) [reply api/main_endpoints/routes/Event.js:18]: along with updating the calls to this function

**Review by allen-tran — APPROVED** [review]
good to me looks

**Pull request merged**